Source/DafnyStandardLibraries/src/DafnyStdLibs/DynamicArray.dfy

```diff
@@ -0,0 +1,244 @@
+    ghost var items : seq<A>
+    ghost var Repr : set<object>
+
+    const defaultValue: A
+    var size: uint32
+    var capacity: uint32
+    var data: array<A>
+
+    const MAX_CAPACITY: uint32 := (TWO_TO_THE_32 - 1) as uint32
+    const MAX_CAPACITY_BEFORE_DOUBLING: uint32 := MAX_CAPACITY / 2
+
+    ghost predicate Valid?()
+      reads this, Repr
+      ensures Valid?() ==> this in Repr
+    {
+      && Repr == {this, data}
+      && capacity != 0
+      && data.Length == capacity as int
+      && size <= capacity
+      && size as int == |items|
+      && items == data[..size]
+    }
+
+    constructor(defaultValue: A, initial_capacity: uint32 := 8)
+      requires initial_capacity > 0
+      ensures size == 0
+      ensures items == []
+      ensures fresh(Repr)
+      ensures Valid?()
+      ensures capacity == initial_capacity
+    {
+      this.defaultValue := defaultValue;
+      items := [];
+      size := 0;
+      capacity := initial_capacity;
+      data := new A[initial_capacity](_ => defaultValue);
+      Repr := {this, data};
+    }
+
+    /**
+    Retrieve the element at index 
+     */
+    function At(index: uint32) : (element: A)
+      reads this, this.data
+      requires index < size
+      requires Valid?()
+      ensures element == data[index] == items[index]
+    {
+      data[index]
+    }
+
+    /**
+    Put element in the array at index 
+     */
+    method Put(index: uint32, element: A)
+      requires index < size
+      requires Valid?()
+      modifies data, `items
+      ensures Valid?()
+      ensures items == old(items)[index := element]
+    {
+      data[index] := element;
+      items := items[index := element];
+    }
+
+    /**
+    Ensure that at least a reserved amount of elements can still be pushed onto the array in constant time
+
+    Returns false only if it was not at all possible to provide the ensurance.
+     */
+    method Ensure(reserved: uint32) returns (s: bool)
+      requires Valid?()
+      ensures Valid?()
+      modifies `capacity, `data, `Repr, data
+      ensures s == (reserved <= MAX_CAPACITY - size)
+      ensures s ==> reserved <= capacity - size
+      ensures var largeEnough := old(reserved <= capacity - size);
+              if !s || largeEnough
+              then unchanged(this) && unchanged(data)
+              else fresh(data)
+    {
+      if reserved > MAX_CAPACITY - size {
+        return false;
+      }
+      if reserved <= capacity - size {
+        return true;
+      }
+      var new_capacity := capacity;
+      while reserved > new_capacity - size
+        decreases MAX_CAPACITY - new_capacity
+        invariant new_capacity >= capacity
+      {
+        new_capacity := DefaultNewCapacity(new_capacity);
+      }
+      Realloc(new_capacity);
+      return true;
+    }
+
+    /**
+    Pop an element from the array in constant time
+     */
+    method PopFast()
+      requires Valid?()
+      requires size > 0
+      modifies `size, `items
```

Comment:
Use `modifies Repr` and some appropriate postconditions about the exported fields.

Same comment for the public methods below.